### PR TITLE
PORTS - Cloudy

### PR DESCRIPTION
**Pull request by CloudyLopez**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | figuring out what went where and then how. and then moving it somewhere else that made sense, and then moving it back.
What was a design decision you made that changed over time over the project? | i thought i was going to work off 3 classes, the reservation tracker, the reservation and the date. i didn't end up using the date. 
What was a concept you gained clarity on or learning that you'd like to share? | do the tests first, like they say, it really does make more sense. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | creating a test where the cost would be correct but making it for 2 days and $400 (200 a day)
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | ensuring that there were only 20 rooms to choose from was an edge case. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | it was both great and awful. it did create more of a succinct idea of how to handle it, but doing the mid review was REALLY helpful and provided way more clarity.

### CODE REVIEW

**Comment by tildeee** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | x
Answer comprehension questions | x
**Design** |
Each class is responsible for a single piece of the program | x, though it might be interesting to pull out the complex date collision into something else... a different method or a different class
Classes are loosely coupled | x
**Wave 1** |
List rooms | x
Reserve a room for a given date range | x
List reservations for a given date | x
Calculate reservation price | x
Invalid date range produces an error | x
Test coverage | you didn't add `SimpleCov` to this project
**Wave 2** |
View available rooms for a given date range | x
Reserving a room that is not available produces an error | in progress (test there, but not implemented yet)
Test coverage | you didn't add `SimpleCov` to this project
**Wave 3** |
Create a block of rooms |
Check if a block has rooms |
Reserve a room from a block |
Test coverage |
**Fundamentals** |
Names variables, classes and modules appropriately | x
Understanding of variable scope - local vs instance | x
Can create complex logical structures utilizing variables | x
Appropriately uses methods to break down tasks into smaller simpler tasks | could practice this more
Understands the differences between class and instance methods | x
Appropriately uses iterators and `Enumerable` methods | there are some opportunities to improve on this
Appropriately writes and utilizes classes | x
Appropriately utilizes modules as a namespace | x
**Wrap Up** |
There is a refactors.txt file | There is no refactors.txt file :(
The file provides a roadmap to future changes | :(
**Additional Feedback** |



Good work on this Cloudy!


Ultimately, the code that is here, I'm actually very happy with: there's a clear understanding of how to define classes, use parameters, difference between local variables and instance variables, and how to test these things. If you had more time to build this project out, this code would be a great foundation.


However, there is one concern I have that feels urgent: there were a few workflow requirements that were missing. You had great git habits, and it's clear that you were doing TDD properly!
1. `rake` didn't run tests for me right away: I had to fix a `require` to `require_relative`, which indicates to me that running tests was different for you
1. `SimpleCov` wasn't configured for this project
1. The `refactors.txt` file that was required at the end of the project isn't there


Otherwise, the comments I have on this project are all opportunities for refactoring! Meaning: I like the code as is, and I want to point out ways that it could be better.


Your code on this project was strong-- with workflow and requirements being unmet, along with the tiny factor that Wave 2 was not quite finished-- bring me to conclude that this is a great submission that almost gets there.

File: lib/reservation.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  class Reservation
+
+    attr_reader :room_number, :checkin_date, :checkout_date, :cost
+
+      def initialize(room_number, checkin_date, checkout_date, cost: 200)
+        # check_date_range(checkin_date, checkout_date)
+        @room_number = room_number
+        @checkin_date = checkin_date
+        if checkout_date < checkin_date
+          raise ArgumentError, "check out cannot happen before check in"
```
Comment (tildeee) [inline lib/reservation.rb:13]: Nice!
Because the nature of raising Errors means that code will halt and stop the program if it reaches it, I might consider refactoring this so it can be a little slimmer...
```ruby
if checkout_date < checkin_date
  raise ArgumentError, "check out cannot happen before check in"
end
```

Mostly, I think you don't need the `else` bit!

File: lib/reservation_tracker.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+module Hotel 
+ 
+    class ReservationTracker
+        attr_reader :reservations, :all_rooms, :all_availabile, :block_reservation
+
+        def initialize(max_rooms)
+            if max_rooms > 20 
+                raise ArgumentError, "this hotel is too large, max 20 rooms"
+            else 
+                @all_rooms = [*1..max_rooms]
```
Comment (tildeee) [inline lib/reservation_tracker.rb:13]: Nice, but is there a requirement to handle a changing number of max rooms? ;) I think that the requirements say that there will always be 20 rooms

File: lib/reservation_tracker.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            @reservations << reservation
+            return reservation
+        end
+
+        def reservations_by_date(date)
+            list_all_rooms = []
+            @reservations.each do |reservation|
+                if date == reservation.checkin_date
+                    list_all_rooms << reservation
+                end
```
Comment (tildeee) [inline lib/reservation_tracker.rb:32]: Nice-- looking back, if you had more time, this `each` loop is a good opportunity to refactor to use an `Enumerable` method (look up `select`)

File: lib/reservation_tracker.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        end
```
Comment (tildeee) [inline lib/reservation_tracker.rb:33]: watch your indentation here-- it's off!

File: lib/reservation_tracker.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        return list_all_rooms
+        end
+       
+        def check_availability(checkin_date, checkout_date)
```
Comment (tildeee) [inline lib/reservation_tracker.rb:37]: It seems like this method is still in progress, but from what I can tell so far: the name `check_availability` isn't my favorite name for this method. What do you want this method to do, and to return? It's unclear from its name... the name `check_availability` seems like it will give back a boolean, but you are doing some logic to add to `booked_rooms`. Just a heads up!

File: lib/room.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+#                 available_room << room
+#             end
+#         return available_rooms
+#         end
+#     end 
+
+
+#     end
+
+# end
```
Comment (tildeee) [inline lib/room.rb:26]: Please feel free to delete your unused files before submission :-)

File: spec/reservation_tracker_spec.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
+require 'date'
+require 'spec_helper'
```
Comment (tildeee) [inline spec/reservation_tracker_spec.rb:2]: My tests wouldn't run unless I changed this to `require_relative` -- How did you get your tests to run?

File: spec/reservation_tracker_spec.rb [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        end
+       
+        it "book rooms is in array for available rooms" do
+            rooms_available = @front_desk.check_availability(@checkin_date2, @checkout_date2)
+            expect(rooms_available).length.must_equal 1
+            expect(rooms_available).length.wont_equal 2
+        end
+
+        it "reservation can starts the day that a different one ends" do
+        @checkin_date4 = Date.new(2019,3,6)
```
Comment (tildeee) [inline spec/reservation_tracker_spec.rb:78]: Indentation!

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -6,3 +6,5 @@
 Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
 
 # Require_relative your lib files here!
+require_relative '../lib/reservation_tracker.rb'
+require_relative '../lib/reservation.rb'
```
Comment (tildeee) [inline spec/spec_helper.rb:10]: You'll want to add the following lines to this file:

```ruby
require 'simplecov'
SimpleCov.start
```
